Adding basic testing for primitive types

vsoch opened the pull request:
Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented on test/source/allocation.cpp:
```diff
@@ -19,13 +19,79 @@ auto get_one(smeagle::Corpus const& corpus, char const* name) {
+    CHECK(parameters[0].location == "%rdi");
+  }
+
+  SUBCASE("void") {
```
Would this be the same if someone wrote "void" as an argument? I know that's not good practice, but we could see it.

hainest replied:
Yes, it would be the same.

hainest commented on test/source/allocation.cpp:
```diff
@@ -19,13 +19,79 @@ auto get_one(smeagle::Corpus const& corpus, char const* name) {
+    REQUIRE(parameters.size() == 0);
+  }
+
+  SUBCASE("one wchar_t") {
```
This case segfaults. It actually brings up a good point: what about Unicode? The AMD64 ABI doc says nothing about it. The C++ standard says only that `wchar_t` should "be large enough to represent any supported character code point". This could be UTF-16 or UTF-32 (aka Unicode). The changes I'm adding to Dyninst will provide support for detecting UTF encodings in DWARF4 or newer. However, I'm not sure what the answer should be for how a register should be allocated. My _guess_ would be the same as a 32-bit int.

hainest commented on test/source/allocation.cpp:
```diff
@@ -19,13 +19,79 @@ auto get_one(smeagle::Corpus const& corpus, char const* name) {
+    CHECK(parameters[0].location == "%rdi");
+  }
+
+  SUBCASE("one float") {
```
We don't correctly handle floating point types yet. I wouldn't have a test case for this- particularly since it should be in %xmm0, not %xmm7.

vsoch replied:
Ah my loop is backwards.

hainest commented on test/source/allocation.cpp:
```diff
@@ -19,13 +19,79 @@ auto get_one(smeagle::Corpus const& corpus, char const* name) {
+    CHECK(parameters[0].location == "%rdi");
+  }
+
+  SUBCASE("one string") {
```
This is an aggregate type (a user-defined class) which we don't handle yet.

vsoch replied:
There's a lot of types we don't handle - right now it just returns two framebases. I'm okay to leave them as templates and modify when we have the implementation! Or at least for the ones I've already written (and then would have to write again).

hainest reviewed (changes requested):
There are a couple of built-in integral types missing: short, long, long long, size_t

We should also test both the explicitly signed and unsigned versions (e.g., char, signed char, unsigned char) as well as the "short hand" variants (e.g., "signed" is really "signed int").

We also want to test all of the fixed-width types from cstdint (https://en.cppreference.com/w/cpp/types/integer).

I wouldn't check the 'type'. That information ultimately comes from the compiler which is allowed to spell "unsigned long long int" as "long int unsigned long" and both are correct. Once I get my changes merged into Dyninst (they're nearly done!), we can generate type names like "Integer64" or "Float32" which is closer to what the model doc has. Those we can definitely check for explicitly.

Thinking ahead a bit, my suggestion on organizing the tests might be to split them up so that each TEST_CASE is one of "integral types", "signed integral types", "unsigned integral types", "float", etc. Then we can add more stuff later like "class", "struct", "array", etc. I would also change the SUBCASE names from "one int" to just "int". I had put the "one X" name there because I was thinking about doing permutation tests. I think that would be best done under a different TEST_CASE.

vsoch commented:
okay all set! https://github.com/buildsi/Smeagle/pull/25/commits/85dce4aa58c47cf57a80026f809af6a78785878c

I suspect many of them we still need to implement (and the answers will need to be changed) but I think we can easily modify the test when the implementation is done. Thank you for the great review! Specifically I've:

 - broken tests into more test cases (and separated them by group in the test file with functions)
 - removed checking of types
 - updated descriptions to not include "one"
 - added fixed types
 - adding missing types
 - reversed generation loop of `%xmm`s - we needed to start at 7 and go down to 0 (since we pop 0 off first).

hainest reviewed (changes requested):
Many of these tests are crashing with a segfault for me, and we're still missing a few signed/unsigned combination. I wrote a quick python script to generate the tests for the integral types and then expanded on the tests you had for floating point and UTF. For some reason, none of the updated tests are crashing anymore- but I only just copied your code! No idea.

https://github.com/hainest/Smeagle/blob/unit_tests/test/source/allocation.cpp
https://github.com/hainest/Smeagle/blob/unit_tests/test/source/libs/allocation.cpp

vsoch commented:
Really? That's weird - they work for me. Did you try running them in the development container?

vsoch commented:
If it's character encodings / newlines or something weird, feel free to update the PR here with the file that works for you.

vsoch commented:
@hainest good to merge here? And do you want to chat about what I can work on next?

hainest commented:
Yup! Let's also add them to the actions here.

>And do you want to chat about what I can work on next?

Surely. I have time this afternoon or anytime tomorrow after 9AM PST. Definitely the next thing to try is the new type information I added to Dyninst to get better type detection in `getRegisterClassFromType`.

vsoch commented:
@hainest I think we are already running them here - scroll down to the test section here: https://github.com/buildsi/Smeagle/pull/25/checks?check_run_id=2881198863. Is that what you mean?

hainest commented:
Oh, awesome! I didn't see they were already there. :+1:

vsoch commented:
okay great! I'll merge here and ping you in a little over an hour - going to go for my late-day-run-oops-its-going-to-thunder-should-have-gone-early-so-many-regrets run :)
